### ISSUE TITLE
Log filled with write plugins failed / back to normal messages

### ISSUE THREAD

**Issue by mnlipp** [collectd/collectd#2823]
*   Version of collectd:
collectd-5.8.0-8.fc28.x86_64
*   Operating system / distribution:
fc28
## Expected behavior
"Silent" operation
## Actual behavior

```
Jun 16 09:59:28 olymp.fritz.box collectd[9192]: Filter subsystem: Built-in target `write': Dispatching value to all write plugins failed with status -1.
Jun 16 09:59:28 olymp.fritz.box collectd[9192]: Filter subsystem: Built-in target `write': Some write plugin is back to normal operation. `write' succeeded.
Jun 16 09:59:38 olymp.fritz.box collectd[9192]: Filter subsystem: Built-in target `write': Dispatching value to all write plugins failed with status -1.
Jun 16 09:59:38 olymp.fritz.box collectd[9192]: Filter subsystem: Built-in target `write': Some write plugin is back to normal operation. `write' succeeded.
Jun 16 09:59:48 olymp.fritz.box collectd[9192]: Filter subsystem: Built-in target `write': Dispatching value to all write plugins failed with status -1.
Jun 16 09:59:48 olymp.fritz.box collectd[9192]: Filter subsystem: Built-in target `write': Some write plugin is back to normal operation. `write' succeeded.
Jun 16 09:59:58 olymp.fritz.box collectd[9192]: Filter subsystem: Built-in target `write': Some write plugin is back to normal operation. `write' succeeded.
```
... and so on.

## Steps to reproduce

*   configure and start collectd ;-)

Here's th: log up to the first failure:
```
Jun 16 10:24:18 olymp.fritz.box systemd[1]: Starting Collectd statistics daemon...
Jun 16 10:24:18 olymp.fritz.box collectd[9536]: plugin_load: plugin "syslog" successfully loaded.
Jun 16 10:24:18 olymp.fritz.box collectd[9536]: plugin_load: plugin "cpu" successfully loaded.
Jun 16 10:24:18 olymp.fritz.box collectd[9536]: plugin_load: plugin "interface" successfully loaded.
Jun 16 10:24:18 olymp.fritz.box collectd[9536]: plugin_load: plugin "load" successfully loaded.
Jun 16 10:24:18 olymp.fritz.box collectd[9536]: plugin_load: plugin "memory" successfully loaded.
Jun 16 10:24:18 olymp.fritz.box collectd[9536]: plugin_load: plugin "apache" successfully loaded.
Jun 16 10:24:18 olymp.fritz.box collectd[9536]: plugin_load: plugin "cpufreq" successfully loaded.
Jun 16 10:24:18 olymp.fritz.box collectd[9536]: plugin_load: plugin "curl_xml" successfully loaded.
Jun 16 10:24:18 olymp.fritz.box collectd[9536]: plugin_load: plugin "df" successfully loaded.
Jun 16 10:24:18 olymp.fritz.box collectd[9536]: plugin_load: plugin "disk" successfully loaded.
Jun 16 10:24:18 olymp.fritz.box collectd[9536]: plugin_load: plugin "hddtemp" successfully loaded.
Jun 16 10:24:18 olymp.fritz.box collectd[9536]: plugin_load: plugin "ipmi" successfully loaded.
Jun 16 10:24:18 olymp.fritz.box collectd[9536]: plugin_load: plugin "virt" successfully loaded.
Jun 16 10:24:18 olymp.fritz.box collectd[9536]: plugin_load: plugin "network" successfully loaded.
Jun 16 10:24:18 olymp.fritz.box collectd[9536]: plugin_load: plugin "ntpd" successfully loaded.
Jun 16 10:24:18 olymp.fritz.box collectd[9536]: plugin_load: plugin "nut" successfully loaded.
Jun 16 10:24:18 olymp.fritz.box collectd[9536]: plugin_load: plugin "ping" successfully loaded.
Jun 16 10:24:18 olymp.fritz.box collectd[9536]: plugin_load: plugin "rrdtool" successfully loaded.
Jun 16 10:24:18 olymp.fritz.box collectd[9536]: plugin_load: plugin "sensors" successfully loaded.
Jun 16 10:24:18 olymp.fritz.box collectd[9536]: syslog: invalid loglevel [debug] defaulting to 'info'
Jun 16 10:24:18 olymp.fritz.box collectd[9536]: plugin_load: plugin "unixsock" successfully loaded.
Jun 16 10:24:18 olymp.fritz.box collectd[9536]: Systemd detected, trying to signal readyness.
Jun 16 10:24:18 olymp.fritz.box systemd[1]: Started Collectd statistics daemon.
Jun 16 10:24:18 olymp.fritz.box audit[1]: SERVICE_START pid=1 uid=0 auid=4294967295 ses=4294967295 subj=system_u:system_r:init_t:s0 msg='unit=collectd comm="systemd" exe="/usr/lib/systemd/s>
Jun 16 10:24:18 olymp.fritz.box collectd[9536]: cpufreq plugin: Found 32 CPUs
Jun 16 10:24:18 olymp.fritz.box collectd[9536]: virt plugin: reader virt-0 initialized
Jun 16 10:24:18 olymp.fritz.box collectd[9536]: Initialization complete, entering read-loop.
Jun 16 10:24:18 olymp.fritz.box collectd[9536]: Init SSL without certificate database
Jun 16 10:24:18 olymp.fritz.box collectd[9536]: Filter subsystem: Built-in target `write': Dispatching value to all write plugins failed with status -1.
Jun 16 10:24:18 olymp.fritz.box collectd[9536]: nut plugin: Connection to (localhost, 3493) established.
Jun 16 10:24:18 olymp.fritz.box collectd[9536]: nut plugin: Connection is unsecured (no SSL).
Jun 16 10:24:18 olymp.fritz.box collectd[9536]: Filter subsystem: Built-in target `write': Some write plugin is back to normal operation. `write' succeeded.
```

**Comment by rpv-tomsk** [comment]
>Steps to reproduce - configure and start collectd ;-)

What is the minimal config to reproduce the issue?

**Comment by mnlipp** [comment]
I've left the `collectd.conf` untouched ([collectd.conf.txt](https://github.com/collectd/collectd/files/2108038/collectd.conf.txt)).

First I added only `rrdtools.conf` in `collect.d`:
```LoadPlugin rrdtool
<Plugin rrdtool>
       DataDir "/var/lib/collectd/rrd"
       CacheTimeout 120
       CacheFlush   900
</Plugin>
```
Results in:
```
Jun 16 12:22:29 olymp.fritz.box systemd[1]: Starting Collectd statistics daemon...
Jun 16 12:22:29 olymp.fritz.box collectd[4714]: plugin_load: plugin "syslog" successfully loaded.
Jun 16 12:22:29 olymp.fritz.box collectd[4714]: plugin_load: plugin "cpu" successfully loaded.
Jun 16 12:22:29 olymp.fritz.box collectd[4714]: plugin_load: plugin "interface" successfully loaded.
Jun 16 12:22:29 olymp.fritz.box collectd[4714]: plugin_load: plugin "load" successfully loaded.
Jun 16 12:22:29 olymp.fritz.box collectd[4714]: plugin_load: plugin "memory" successfully loaded.
Jun 16 12:22:29 olymp.fritz.box collectd[4714]: plugin_load: plugin "apache" successfully loaded.
Jun 16 12:22:29 olymp.fritz.box collectd[4714]: plugin_load: plugin "rrdtool" successfully loaded.
Jun 16 12:22:29 olymp.fritz.box collectd[4714]: Systemd detected, trying to signal readyness.
Jun 16 12:22:29 olymp.fritz.box systemd[1]: Started Collectd statistics daemon.
Jun 16 12:22:29 olymp.fritz.box collectd[4714]: Initialization complete, entering read-loop.

```
Works without problems.

Now I have re-enabled all my `*.conf` files one by one. Obviously, the culprit is chrony:
```
LoadPlugin chrony

<Plugin chrony>
       Host    "localhost"
       Port    "323"
       Timeout "2"
</Plugin>
```
When I add this, the problem starts.

**Comment by rpv-tomsk** [comment]
Please provide clean minimal config, without `Include "/etc/collectd.d"` and comments.

**Comment by rpv-tomsk** [comment]
Hi!

I added some messages output for error cases, which was silent before.
When #2825 gets merged and CI builds new packages (automatically), can you please try package from https://pkg.ci.collectd.org/rpm/master/  ? Also you can build Collectd from sources and try to find a cause.

**Comment by mnlipp** [comment]
Okay, thanks a lot, but don't expect any more feedback soon. Using the `collectd.spec` from fc28, the code from master doesn't build anymore because of `configure: error: strtok_r is not available`. Running `./build.sh` and `./configure` does work, though, so it's not a collectd issue. Obviously, the `rpmbuild` sets some `*FLAGS` in a way that makes building impossible. Which one exactly triggers the problem is beyond my (short-term) understanding. 

I know that I could simply install collectd manually, i.e. without building packages first, but because I plan to use the server that I'm configuring for a long time, I want to keep things clean. So, I'll wait for the Fedora package maintainer to provide a new version with an updated `collectd.spec`.

PS: Just in case you're interested, here's what rpmbuild does to the environment and the project before configuring:
```
+ CFLAGS='-O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection'
+ export CFLAGS
+ CXXFLAGS='-O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection'
+ export CXXFLAGS
+ FFLAGS='-O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -I/usr/lib64/gfortran/modules'
+ export FFLAGS
+ FCFLAGS='-O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -I/usr/lib64/gfortran/modules'
+ export FCFLAGS
+ LDFLAGS='-Wl,-z,relro  -Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld'
+ export LDFLAGS
+ '[' 1 = 1 ']'
+++ dirname ./configure
++ find . -name config.guess -o -name config.sub
+ for i in $(find $(dirname ./configure) -name config.guess -o -name config.sub)
++ basename ./build-aux/config.sub
+ '[' -f /usr/lib/rpm/redhat/config.sub ']'
+ /usr/bin/rm -f ./build-aux/config.sub
++ basename ./build-aux/config.sub
+ /usr/bin/cp -fv /usr/lib/rpm/redhat/config.sub ./build-aux/config.sub
'/usr/lib/rpm/redhat/config.sub' -> './build-aux/config.sub'
+ for i in $(find $(dirname ./configure) -name config.guess -o -name config.sub)
++ basename ./build-aux/config.guess
+ '[' -f /usr/lib/rpm/redhat/config.guess ']'
+ /usr/bin/rm -f ./build-aux/config.guess
++ basename ./build-aux/config.guess
+ /usr/bin/cp -fv /usr/lib/rpm/redhat/config.guess ./build-aux/config.guess
'/usr/lib/rpm/redhat/config.guess' -> './build-aux/config.guess'
+ '[' 1 = 1 ']'
+ '[' x '!=' 'x-Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld' ']'
++ find . -name ltmain.sh
+ for i in $(find . -name ltmain.sh)
+ /usr/bin/sed -i.backup -e 's~compiler_flags=$~compiler_flags="-Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld"~' ./build-aux/ltmain.sh
```

**Comment by rubenk** [comment]
> So, I'll wait for the Fedora package maintainer to provide a new version with an updated collectd.spec.

As Fedora package maintainer, that's going to take a while since collectd first has to do a new release and we don't have a date planned yet.

can you share your config.log? That should say why strtok_r could not be detected.

**Comment by mnlipp** [comment]
Sure. Here you are: 
[config.log](https://github.com/collectd/collectd/files/2114534/config.log)

**Comment by rubenk** [comment]
Thanks, the relevant part is this:
```
configure:20328: checking for strtok_r
configure:20365: gcc -o conftest -Wall -Werror  -Wl,-z,relro  -Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld conftest.c  >&5
/usr/bin/ld: /tmp/cc90PsQM.o: relocation R_X86_64_32 against `.rodata' can not be used when making a shared object; recompile with -fPIC
/usr/bin/ld: final link failed: Nonrepresentable section on output
collect2: error: ld returned 1 exit status
configure:20365: $? = 1
```

No idea what's going on there. Not being able to see what flags the linker uses makes it even harder.
What does your HEAD point to? Do you have f9aaf837d506abb28e607ff86da03cbdf00cb40d in your working copy?

**Comment by rubenk** [comment]
Can you check if 9a065c51cae5997052e5cfbb7e41f48bebef2167 helps? I just pushed that to master.

**Comment by mnlipp** [comment]
With the current master, I can successfully complete the configure step. However, the Fedora spec does not seem to be up-to-date with respect to dependencies or some flags are not handled properly. Building stops with `configure: error: "Some plugins are missing dependencies - see the summary above for details"`.

I'll have a closer look at those. Here's an example: `configure` reports: `libaquaero5 . . . . . no (libaquaero5.h not found)`, but `configure` was called with `--disable-aquaero`. Therefore I'd expect `(disabled on command line)` rather than the "header not found".

**Comment by mnlipp** [comment]
(Okay. I've understood by now that checking for the libraries happens independent of the configured plugins.) 

I succeeded in building the packages after disabling ampq1. It wasn't disabled in the Fedora spec file and since it is also not available as package in the Fedora repository, I assume that it was introduced after the 5.8.0 release.

Anyway, my packages, built from master, don't exhibit the problem any more. So whatever the cause was, it has been fixed between the 5.8.0 release and now.

**Comment by rubenk** [comment]
> (Okay. I've understood by now that checking for the libraries happens independent of the configured plugins.)

Yes, that's a bit confusing and tripped me up a few times as well. Ideally we wouldn't check for the library if the plugin is explicitly disabled. Unfortunately fixing that will require rewrite half of configure.ac..

> 
> I succeeded in building the packages after disabling ampq1. It wasn't disabled in the Fedora spec file and since it is also not available as package in the Fedora repository, I assume that it was introduced after the 5.8.0 release.

Correct, the amqp1 plugin was only just introduced in master.
> 
> Anyway, my packages, built from master, don't exhibit the problem any more. So whatever the cause was, it has been fixed between the 5.8.0 release and now.

That's good to hear. It would be great if you could bisect this but if you don't have the time to do that I understand.
